### PR TITLE
[Reader] #1983 Render police: comment indicator

### PR DESCRIPTION
**Pull request by NickHeiner**
Connect #1983.

Branched off #2073. 

We are eliminating another set of `O(n)` renders from both the good and wasted renders. These results are generated by filtering the document list with a search query.

To fix the issues, move logic from `utils` into `selectors`, so equivalent inputs will give shallowly equal outputs.

### Good Renders
image

### Wasted Renders
image

### Good Renders
image

### Wasted Renders
image

### CODE REVIEW

File: client/app/reader/PdfViewer.jsx [hunk 1]
```diff
@@ -139,7 +138,7 @@ export class PdfViewer extends React.Component {
 const mapStateToProps = (state, ownProps) => ({
   documents: getFilteredDocuments(state),
   editingCommentsForCurrentDoc:
-    _.some(getAnnotationByDocumentId(state, Number(ownProps.match.params.docId)), 'editing'),
+    _.some(makeGetAnnotationsByDocumentId(state)(Number(ownProps.match.params.docId), 'editing')),
```
Comment (mdbenjam) [inline client/app/reader/PdfViewer.jsx]: I think the closing parens are off here. 'editing' should be part of the some call.

File: client/app/reader/selectors.js [hunk 2]
```diff
@@ -11,3 +11,40 @@ export const getFilteredDocuments = createSelector(
...
+
+const getEditingAnnotations = (state) => state.editingAnnotations;
+const getPendingEditingAnnotations = (state) => state.ui.pendingEditingAnnotations;
+const getAnnotations = (state) => state.annotations;
+const getPendingAnnotations = (state) => state.ui.pendingAnnotations;
+
+export const makeGetAnnotationsByDocumentId = createSelector(
+  [getEditingAnnotations, getPendingEditingAnnotations, getAnnotations, getPendingAnnotations],
+  (editingAnnotations, pendingEditingAnnotations, annotations, pendingAnnotations) =>
+    _.memoize(
```
Comment (mdbenjam) [inline client/app/reader/selectors.js]: It seems a little weird to use a _.memoize function inside of a selector. What about making this a set of nested selectors?

File: client/app/reader/selectors.js [hunk 3]
```diff
@@ -11,3 +11,40 @@ export const getFilteredDocuments = createSelector(
...
+        uniqBy('id').
+        reject('pendingDeletion').
+        filter({ documentId: docId }).
+        value()
+    )
+);
+
+export const getAnnotationsPerDocument = createSelector(
+  [getFilteredDocuments, makeGetAnnotationsByDocumentId],
+  (documents, getAnnotationsByDocumentId) =>
```
Comment (mdbenjam) [inline client/app/reader/selectors.js]: Maybe I'm missing something simple, but I'm really confused where `getAnnotationsByDocumentId` is coming from? Don't you need to call `makeGetAnnotationsByDocumentId` with the state to get this function?

File: client/test/karma/reader/selectors-test.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            'temp-guid': {
+              id: 'temp-guid',
+              comment: 'pending annotation',
+              documentId
+            }
+          }
+        }
+      };
+
+      expect(makeGetAnnotationsByDocumentId(state)(documentId)).to.deep.equal([
```
Comment (mdbenjam) [inline client/test/karma/reader/selectors-test.js]: Is there any use to testing the memoization of the selector? To make sure it memoizes when we expect it to? Maybe that's erring too far into testing details for react-select.

File: client/app/reader/PdfListView.jsx [hunk 5]
```diff
@@ -380,7 +333,7 @@ export class PdfListView extends React.Component {
           className="document-list-header-comments">
           Comments
         </div>,
-        valueFunction: (doc) => <ConnectedCommentIndicator doc={doc} />
+        valueFunction: (doc) => <CommentIndicator docId={doc.id} />
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: Honestly I'm not totally sure. I feel like I had a reason but I'm not able to remember it now. I can change it back if you prefer.

File: client/app/reader/selectors.js (hunk 2 above)
Comment (NickHeiner) [inline client/app/reader/selectors.js]: I got the idea to memoize it from here: https://github.com/reactjs/reselect#q-how-do-i-create-a-selector-that-takes-an-argument.

What do you mean by nested selectors? I'm happy for an alternative approach.

File: client/app/reader/selectors.js (hunk 3 above)
Comment (NickHeiner) [inline client/app/reader/selectors.js]: When you have the following:

```js
const fooSelector = (state) => state.foo;
createSelector(
  [fooSelector],
  (foo) => foo.bar;
```

`createSelector` will call `fooSelector(state)` and make the result available as `foo` to the selector function we're passing. 

In this case, `makeGetAnnotationsByDocumentId` is called with `state`, and the result is `getAnnotationsByDocumentId`.

Does that make sense? I'm happy to chat through it synchronously.

**Pull request merged**